cmd/lifecycle/builder.go

```diff
@@ -48,26 +60,160 @@ func (b *buildCmd) Args(nargs int, args []string) error {
 }
 
 func (b *buildCmd) Privileges() error {
-	// builder should never be run with privileges
-	if priv.IsPrivileged() {
+	_, privGroup, _, err := b.readData()
+	if err != nil {
+		return err
+	}
+
+	hasPrivGroups := len(privGroup.Group) > 0
+
+	// builder should never be run with privileges if there aren't any stack buildpacks
+	if !hasPrivGroups && priv.IsPrivileged() {
 		return cmd.FailErr(errors.New("refusing to run as root"), "build")
+	} else if hasPrivGroups && !priv.IsPrivileged() {
+		return cmd.FailErr(errors.New("must run as root"), "build")
 	}
 	return nil
 }
 
 func (b *buildCmd) Exec() error {
-	group, plan, err := b.readData()
+	group, privGroup, plan, err := b.readData()
+	if err != nil {
+		return err
+	}
+
+	if err := verifyBuildpackApis(group, privGroup); err != nil {
+		return err
+	}
+
+	if len(privGroup.Group) == 0 {
+		return b.build(group, plan)
+	}
+	return b.buildWithReexec(group, privGroup, plan)
+}
+
+func (ba buildArgs) buildAll(group, privGroup lifecycle.BuildpackGroup, plan lifecycle.BuildPlan, rootuid, rootgid, uid, gid int) error {
+	if len(privGroup.Group) > 0 {
+		if err := priv.RunAsEffective(rootuid, rootgid); err != nil {
+			return err
+		}
+		if err := ba.stackBuild(privGroup, plan); err != nil {
+			return err
+		}
+		if err := priv.RunAsEffective(uid, gid); err != nil {
+			return err
+		}
+	}
+
+	return ba.build(group, plan)
+}
+
+func (ba buildArgs) stackBuild(privGroup lifecycle.BuildpackGroup, plan lifecycle.BuildPlan) error {
+	stackSnapshotter, err := snapshot.NewKanikoSnapshotter("/", ba.layersDir, ba.platformDir)
 	if err != nil {
 		return err
 	}
-	if err := verifyBuildpackApis(group); err != nil {
+
+	workspaceDir, err := ioutil.TempDir("", "stack-workspace")
+	if err != nil {
+		return err
+	}
+
+	builder, err := ba.createBuilder(privGroup, plan)
+	if err != nil {
 		return err
 	}
-	return b.build(group, plan)
+
+	builder.Snapshotter = stackSnapshotter
+	builder.AppDir = workspaceDir
+	builder.BuildpacksDir = ba.stackBuildpacksDir
+
+	_, err = ba.execBuild(builder)
+	return err
 }
 
 func (ba buildArgs) build(group lifecycle.BuildpackGroup, plan lifecycle.BuildPlan) error {
-	builder := &lifecycle.Builder{
+	builder, err := ba.createBuilder(group, plan)
+	if err != nil {
+		return err
+	}
+	md, err := ba.execBuild(builder)
+	if err != nil {
+		return err
+	}
+
+	if err := lifecycle.WriteTOML(launch.GetMetadataFilePath(ba.layersDir), md); err != nil {
+		return cmd.FailErr(err, "write build metadata")
+	}
+	return nil
+}
+
+func (ba buildArgs) buildWithReexec(group, privGroup lifecycle.BuildpackGroup, plan lifecycle.BuildPlan) error {
+	err := ba.stackBuild(privGroup, plan)
+	if err != nil {
+		return err
+	}
+
+	if len(group.Group) > 0 {
+		if err = ba.buildAsSubProcess(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ba buildArgs) buildAsSubProcess() error {
+	if err := priv.RunAs(ba.uid, ba.gid); err != nil {
```

Comment:
I know there was conversation at some point (though I can't find it) around whether a "stack build" should share the same `Exec` as a regular build. It feels weird to me that we're calling `priv.RunAs` here whereas everywhere else in the code it's being called in `Privileges()`. It makes me wonder if we need to pull the two flows apart at a higher level of abstraction.

Reply:
This part is a bit strange. I'd be down for figuring out a cleaner way to go about this. Ideally without re-exec entirely. That bit is gross anyway. I'm not convinced it is buying us anything over only dropping privileges.